Generate TypeScript typings for functions supplied to dnsconfig.js

j-f1 opened the pull request:
A `.d.ts` file with full documentation is now generated from existing docs files (with a few manual additions). If placed next to a `dnsconfig.js` file, supporting editors (VS Code but also others as well) will have a significantly improved DX.

I’m marking this as WIP because I need to do a bunch more, but I’m very interested in feedback!
- [x] find a better front matter parsing library (the current one behaves strangely when `---` appears later on in the document)
- [x] manually review the generated typings for inaccuracies when compared with the prose docs
- [x] Fix typings for `FETCH` (currently it uses `fetch`’s typings which are only available to TS code that enables the DOM library)
- [x] come up with a way to ship the generated file
  - one option would be to bundle it with the program (ensuring that it always provides the correct typings that correspond to the available functionality) via a new command that spits out a `dnscontrol.d.ts` file into the current directory
  - another would be to host the files somewhere (maybe check the generated `types.d.ts` file into Git so that it can be fetched from the current release’s tag?)
  - maybe add a way to generate a suitable `jsconfig.json` to configure TypeScript in editors? 
- [x] Figure out if there’s a way to make `CLI_DEFAULTS` cause the names it’s passed to show up as variables **(Result: Probably not)**
- [x] Docs: How to add new record types and get correct TS typings
- [x] Docs: How to use TS with DNSControl

Documentation is shown inline:
<img width="735" alt="Screenshot_2022-11-29 15 37 36" src="https://user-images.githubusercontent.com/25517624/204642641-5914c28e-3d77-43d5-8dea-a2fef0332ad0.png">

Autocomplete shows all available record types:
<img width="650" alt="Screenshot_2022-11-29 15 39 40" src="https://user-images.githubusercontent.com/25517624/204642864-945fbf65-47e7-42a8-86d3-1b0bc7c4e2e1.png">

tlimoncelli reviewed (commented):
Holy cow!  I didn't even know that kind of thing was possible!

This is awesome!  Yes, please and thank you!

j-f1 commented:
Some questions (actually 1 question) based on my reading through the docs:

- For [`SSHFP`](https://dnscontrol.org/js#SSHFP), arę the reserved “0” algorithm/type supported as values? Or should it just be 1/2/3 allowed as parameters? (also, it would be a neat addition to support the string values like `"DSA"` too)

tlimoncelli commented:
> Some questions (actually 1 question) based on my reading through the docs:
> 
> * For [`SSHFP`](https://dnscontrol.org/js#SSHFP), arę the reserved “0” algorithm/type supported as values? Or should it just be 1/2/3 allowed as parameters? (also, it would be a neat addition to support the string values like `"DSA"` too)

According to [the source of all knowledge](https://en.wikipedia.org/wiki/SSHFP_record#) 0 is reserved. I guess we could play it safe and permit it.  Though, the world won't end if we guess wrong as I don't think many people use SSHFP ;-)

Tom

j-f1 commented:
Ok, I think this is ready for review! I’ve added a `dnscontrol types` command that downloads the version of the `.d.ts` file corresponding to the current build of `dnscontrol` from GitHub. I definitely want to add some more documentation but before I do that I want to make sure this is a good approach. (I’d also love any feedback on the code since this is my first time writing Go!)

tlimoncelli commented on commands/generateTypes.go:
```diff
@@ -0,0 +1,73 @@
+var _ = cmd(catUtils, func() *cli.Command {
+	var args TypesArgs
+	return &cli.Command{
+		Name:  "types",
```
I'm not in love with having the subcommand called "types". However if that's what other similar systems use, I'd rather be consistent with them.  Could the name refer to typescript (or ts) to be clear what it is for?

j-f1 replied:
The key in the JS `package.json` is `types` (which I kind of based it on). I kind of shy away from using `typescript` since I want it to be clear that this doesn’t actually *run* TypeScript, it just emits the file. Maybe `dnscontrol export-ts-types` or `emit-ts-types` or something?

tlimoncelli replied:
How about `write-types`?  Its short and easy(ish) to remember.

tlimoncelli reviewed (commented):
I only did a shallow review of the Go code. I'll do a deeper look soon.

Two more notes:
1) Documentation: Add docs about how to use, etc.
2) Release engineering: Document what needs to be done to generate the file and any other release engineering steps (i.e. is there a "go generate" or equivalent that needs to run?)

j-f1 commented on build/build.go:
```diff
@@ -16,7 +16,7 @@ var goos = flag.String("os", "", "OS to build (linux, windows, or darwin) Defaul
 func main() {
 	flag.Parse()
-	flags := fmt.Sprintf(`-s -w -X main.SHA="%s" -X main.BuildTime=%d`, getVersion(), time.Now().Unix())
+	flags := fmt.Sprintf(`-s -w -X "main.SHA=%s" -X main.BuildTime=%d`, getVersion(), time.Now().Unix())
```
Yes. Previously, the SHA ended up being like `"abc123"` instead of `abc123`. (not sure why, it’s weird that the quotes aren’t properly handled by the shell)

j-f1 commented:
- docs: added to TODO above
- release engineering: added! you just need to do `go generate` (which _should_ be run periodically anyway to update the matrix)

cafferata commented on docs/release-engineering.md:
````diff
@@ -12,10 +12,23 @@ CircleCI will do most of the work for you. You will need to edit the draft relea
 
 ```shell
 git checkout master
+go generate
````
I make a note to myself to do one last `git pull`. Maybe you also think this is a good idea and can you include it as a necessary step?

```suggestion
git pull
go generate
```

cafferata commented on build/generate/functionTypes.go:
```diff
@@ -0,0 +1,232 @@
+			}
+
+			body = body + "\n"
+			body = strings.ReplaceAll(body, "{{site.github.url}}", "https://www.dnscontrol.org/")
```
After @tlimoncelli has undone the Github custom domain configuration https://github.com/StackExchange/dnscontrol/issues/1843#issuecomment-1337828223, these (links) no longer work.

cafferata commented on docs/_functions/domain/SRV.md:
```diff
@@ -7,6 +7,9 @@ parameters:
   - port
   - target
   - modifiers...
+parameter_types:
+  weight: number
+  port: number
```
Let's face it, I'm not familiar with TypeScript typings (configurations). So I was wondering, when do you include which `parameter_types`? I feel like I'm missing `priority` here?

```suggestion
parameter_types:
  priority: number
  weight: number
  port: number
```

j-f1 replied:
I set a bunch of defaults here: https://github.com/StackExchange/dnscontrol/pull/1830/files#diff-76e760c4509933f6592ab7a1e50b4934736913b3cdfd91c00e1232ff33a81e82R52-R62. That said, it might make sense to get rid of that and explicitly specify all the parameter types for every function. Thoughts?

cafferata replied:
Aha! I think I personally would opt for the duplication;
1. Everything in one place
2. No discussion when something is or isn't 'defaults'.
3. No need to look back on a PR if the missing param happens to be defined as a default.
4. Defaults could be left behind during refactoring activities.

cafferata commented:
In response to commit 'Remove default parameter types' https://github.com/StackExchange/dnscontrol/pull/1830/commits/7d6de8b9d5d408252d3f90735eb75bebfa40a8c4, I now see the knowledge of a function being recorded in three places. Perhaps it is easier to talk in an example situation. 

Let's take [SPF_BUILDER](https://stackexchange.github.io/dnscontrol/js#SPF_BUILDER) as an example. The knowledge about the `parameters` is currently recorded in:
1. `parameters` which are shown in the header.

    <img width="1280" alt="dnscontrol-function-header" src="https://user-images.githubusercontent.com/1150425/205858539-59dfa53b-f113-4af0-a480-2e1825e07bc3.png">

2. `parameter_types` which is not shown, is only used for `TypeScript typings`.

    <img width="1280" alt="dnscontrol-text-documentation" src="https://user-images.githubusercontent.com/1150425/205858656-0a1e0de0-c3a3-4231-9db1-948aecae36f9.png">

3. `The parameters` section, which is described textually in the documentation page.

I don't want to make your pull request unnecessarily bigger. I did want to express my concerns about this error-proneness that this can create.

I'm leaning towards a solution where only the new `parameter_types` is used. My documentation pull request https://github.com/StackExchange/dnscontrol/pull/1809 can then use this. In which the header display (_screenshot 1_) is removed and more work is done towards an elaboration as can be seen in screenshot 2. With possibly a textual addition as is now done within `SPF_BUILDER`.

What do you guys think @j-f1, @tlimoncelli?

tlimoncelli commented:
Some feedback:

1. This is awesome. I'm no VS Code expert but I typed "dnscontrol types" and then started VS Code and suddenly typing `D(` autocompleted! Wow!
2. We need to document this really well.  Assume someone with no VS Code (or other) experience wants to use this. How do they get started?  What do they need to do to update this (if anything)?
3. I don't like the fact that having `dnsconfig.js` and `dnscontrol.d.ts` in the same directory means bash autocompletion stops at `dnscon` when it used to go all the way to `dnsconfig.js`.  (is there an emoji for "crying like a baby"?)

Looking forward to seeing this in production!

tlimoncelli commented:
Cool cool.

The points in (2) should be put in the docs somewhere.  Please add a new page in `$git/dnscontrol/docs` and link to it on the first page.

Point (3): I like `types-dnscontrol.d.ts`.  Does this file need to be in the same directory as `dnsconfig.js`?

j-f1 commented:
> Does this file need to be in the same directory as `dnsconfig.js`?

Yes and no. It can’t be outside of the directory `dnsconfig.js` is in, but it can be in a subdirectory.

tlimoncelli commented on docs/typescript.md:
```diff
@@ -0,0 +1,27 @@
+layout: default
+title: Using TypeScript with DNSControl
+---
+
```
Please add a note that: This is is an experimental feature and may change.

tlimoncelli reviewed (commented):
Just a few small comments and I think we're ready to go.  I'm asking that we mark this feature as "experimental" so that we can be empowered to iterate on it as we get feedback.

tlimoncelli commented:
Great!  Thanks!

tlimoncelli commented:
I'm merging this as-is.   In the release notes I'll point out that it is experimental and ask people to submit issues if they have suggestions and/or improvements.  Once the comments die down we'll make it non-experimental.

I'm super excited about this change. I can't wait to show it to my coworkers!

cafferata commented on docs/_functions/global/FETCH.md:
```diff
@@ -3,6 +3,8 @@ name: FETCH
   - url
   - args
+ts_ignore: true
+# Make sure to update fetch.d.ts if changing the docs below!
```
@j-f1 I've made some changes from https://github.com/StackExchange/dnscontrol/pull/1809, how do I generate the new `fetch.d.ts? `go generate` only generates `types/dnscontrol.d.ts`.

j-f1 replied:
Copy+paste, unfortunately.

cafferata replied:
Perhaps it is a good idea to ensure this necessary step in the docs? https://docs.dnscontrol.org/release/release-engineering#step-1.-rebuild-generated-files

j-f1 replied:
What do you mean? Unless someone changes the `FETCH` function’s interface without updating the docs (which is unlikely), they’ll see this message.

cafferata replied:
Thanks for your comment. I'll further explain my comment with a code example. From https://github.com/StackExchange/dnscontrol/commit/55889582aeddc92d424a50f4577be030dfd04f49#diff-bda8825028c14bd7feb8eb5463e708023019b3ff7f710af883e4b63eba0b9ec3R12 I had to change the URLs. By copy/paste `global/FETCH.md` it's generates this diff which (*tested with JetBrains IDE*) is not valid.

```diff
- Compared to `fetch` from Fetch API, `FETCH` will call [PANIC](https://dnscontrol.org/js#PANIC) to terminate the execution of the script, and therefore DNSControl, if a network error occurs.
+ Compared to `fetch` from Fetch API, `FETCH` will call [PANIC](PANIC.md) to terminate the execution of the script, and therefore DNSControl, if a network error occurs.
```

This should be (*I guess*) a full URL reference?

```diff
- Compared to `fetch` from Fetch API, `FETCH` will call [PANIC](https://dnscontrol.org/js#PANIC) to terminate the execution of the script, and therefore DNSControl, if a network error occurs.
+ Compared to `fetch` from Fetch API, `FETCH` will call [PANIC](https://docs.dnscontrol.org/language-reference/top-level-functions/panic) to terminate the execution of the script, and therefore DNSControl, if a network error occurs.
```

I suspect you corrected this URL once? So the previous answer 'copy/paste' doesn't seem to be entirely valid. Or maybe I'm overlooking something?

j-f1 replied:
Ah yeah, great point. I must have changed those URLs, then. I’m not convinced it’s worth the effort to thoroughly document this since you’re likely to be the last person to touch the file (and especially the links) for a long time.
